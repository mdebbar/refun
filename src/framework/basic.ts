// Closure Compiler doesn't like TS enums, so we have to use a class with static
// fields to emulate an enum.
export class DirtyState {
  static Clean = 1;
  static ChildrenOnly = 2;
  static SelfAndChildren = 3;
}
