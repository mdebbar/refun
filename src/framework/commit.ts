import { AppNode, DirtyState, myNode } from './core';
import { SingleUI } from './core/components';
import { trackSkips } from '../instrumentation';

export type Commit = any;
// TODO[types]: This is useful, please bring it back!
export type CommitUI<C extends Committer> = SingleUI;

const commitTracker = trackSkips('commit', 1000);

// A `Node` in the commit tree. It knows how to commit itself during the commit
// phase after the build is complete. In the case of a browser, committing means
// updating html attributes, appending children, applying styles, etc.
export abstract class Committer<D extends Commit = Commit> {
  dirty: DirtyState = DirtyState.Clean;
  private lastCommit: D | null = null;

  needsCommit() {
    this.dirty = DirtyState.SelfAndChildren;
  }

  needsToCheckChildCommits() {
    if (this.dirty === DirtyState.Clean) {
      this.dirty = DirtyState.ChildrenOnly;
    }
  }

  commit(node: CommitNode<this, any>): D {
    if (this.dirty === DirtyState.SelfAndChildren) {
      commitTracker && commitTracker.hit();
    } else {
      commitTracker && commitTracker.skip();
    }

    if (this.dirty === DirtyState.Clean) {
      return this.lastCommit as D;
    }

    const childCommits = commitChildrenRecursive(node);
    const newCommit =
      this.dirty === DirtyState.ChildrenOnly
        ? (this.lastCommit as D)
        : this.commitSelf(this.lastCommit, childCommits);
    this.diffChildren(newCommit, childCommits);
    this.lastCommit = newCommit;
    this.dirty = DirtyState.Clean;
    return newCommit;
  }

  abstract commitSelf(lastCommit: D | null, newChildren: D[]): D;

  abstract diffChildren(newCommit: D, newChildren: D[]): void;
}

export class NoopCommitter extends Committer<void> {
  commitSelf() {}
  diffChildren() {}
}

export class CommitNode<C extends Committer<Commit>, S> extends AppNode<S> {
  constructor(initialState: S, committer: C) {
    super(initialState);
    this.committer = committer;
  }

  committer: C;

  needsChildrenBuild() {
    super.needsChildrenBuild();
    this.committer.needsToCheckChildCommits();
  }

  protected postBuild() {
    super.postBuild();
    this.committer.commit(this);
  }
}

// Commits the children of `node` recursively. As it walks down the tree, it'll
// add children to `arr`.
//
// QUESTION: should we allow commit nodes to throw? e.g. when adding children to
//           an <img>.
function commitChildrenRecursive<D extends Commit>(
  node: AppNode<any> | CommitNode<Committer<D>, any>,
  arr: D[] = [],
): D[] {
  for (let i = 0; i < node.children.length; i++) {
    const child = node.children[i];
    if (child != null) {
      if (isCommitNode<D>(child)) {
        // When a child is committable, we commit it recursively and add its
        // commit as a child to the ancestor commit.
        const result = child.committer.commit(child);
        if (!result) {
          throw new Error('Expected commit to return something.');
        }
        arr.push(result);
      } else {
        // Continue recursively in the tree to find more commit nodes.
        commitChildrenRecursive(child, arr);
      }
    }
  }
  return arr;
}

function isCommitNode<D>(
  node: AppNode<any>,
): node is CommitNode<Committer<D>, any> {
  return node instanceof CommitNode;
}

export function myCommitNode<C extends Committer<any>>(
  createCommitter: () => C,
) {
  return myNode(() => new CommitNode(null, createCommitter()));
}
