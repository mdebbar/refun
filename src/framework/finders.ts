import { AppNode, Component } from './core';
import { COMPONENT_TYPE } from './core/components';

type Finder = (node: AppNode<any>) => boolean;

export function findAncestor(
  node: AppNode<any>,
  cb: Finder,
): AppNode<any> | null {
  let current: AppNode<any> | null = node;
  while (current != null) {
    if (cb(current)) {
      return current;
    }
    current = current.parent;
  }
  return null;
}

export function findDescendant(node: AppNode<any>, cb: Finder): AppNode<any> {
  if (cb(node)) {
    return node;
  }

  for (let i = 0; i < node.children.length; i++) {
    const child = node.children[i];
    if (child != null) {
      const found = findDescendant(child, cb);
      if (found) return found;
    }
  }

  throw new Error('Could not find the requested node');
}

export function findDescendantOfType(
  node: AppNode<any>,
  component: Component<any>,
): AppNode<any> {
  const type = component()[COMPONENT_TYPE];
  return findDescendant(node, n => n.ui[COMPONENT_TYPE] === type);
}
