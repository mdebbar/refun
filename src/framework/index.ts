export * from './core';
export { committer, CommitNode, Committer, NoopCommitter } from './commit';
export { run } from './run';
// export * from './finders';
export { sleep, animationFrame, animated, gen } from './suspense';
