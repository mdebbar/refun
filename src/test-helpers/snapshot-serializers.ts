import { AppNode } from '../framework';

type any2s = (value: any) => string;
type s2s = (str: string) => string;

export const stateSerializer = {
  test(value: any) {
    return value instanceof AppNode;
  },

  print(value: AppNode<any>, serialize: any2s, indent: s2s) {
    return [
      `%${value.a__name}%`,
      serialize(value.state),
      '[',
      indent(value.children.map(serialize).join(',\n')),
      ']',
    ].join('\n');
  },
};
