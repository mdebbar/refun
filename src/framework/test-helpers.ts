import { Committer } from './commit';

let incrementalCommitValue = 0;

// This function has to be called at the beginning of any test that uses
// jest snapshots with commits.
export function resetCommitValue() {
  incrementalCommitValue = 0;
}

export class TestCommit {
  // Can be used to check the order of commit creation and application.
  value: number;
  children: TestCommit[];

  constructor(value: number) {
    this.value = value;
    this.children = [];
  }
}

export class TestCommitter extends Committer<TestCommit> {
  commitSelf(lastCommit: TestCommit, newChildren: TestCommit[]): TestCommit {
    if (!lastCommit) {
      return new TestCommit(incrementalCommitValue++);
    }
    lastCommit.value = incrementalCommitValue++;
    return lastCommit;
  }

  diffChildren(newCommit: TestCommit, newChildren: TestCommit[]): void {
    newCommit.children = newChildren;
  }
}
