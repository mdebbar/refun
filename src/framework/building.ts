import { DirtyState } from './basic';
import { commit, CommitNode } from './committing';
import { Render, SingleUI, UI } from './types';
import { SkipRate } from '../instrumentation';

const rate = new SkipRate('build', 0);

// Builds the tree starting at `node` while respecting the `dirtyState` flag.
function build(node: AppNode) {
  switch (node.dirtyState) {
    case DirtyState.Clean:
      // If the node is clean, skip the whole subtree.
      rate.total();
      rate.skip();
      break;

    case DirtyState.ChildrenOnly:
      rate.total();
      rate.skip();
      buildChildren(node);
      break;

    case DirtyState.SelfAndChildren:
      forceBuild(node);
      break;
  }
}

function buildChildren(node: AppNode) {
  const children = node.children;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (child) {
      build(child);
    }
  }
}

function forceBuild(node: AppNode) {
  rate.total();
  const children = node.render(node);
  node.dirtyState = DirtyState.Clean;

  // TODO[perf]: Reuse nodes instead of throwing them away. When a node is
  //             recycled, its state needs to be reset.
  if (Array.isArray(children)) {
    const count = buildMultiUI(children, node, 0);
    if (node.children.length > count) {
      node.children.length = count;
    }
  } else {
    buildUIAtIndex(children, node, 0);
    if (node.children.length > 1) {
      node.children.length = 1;
    }
  }
}

function buildUIAtIndex(ui: SingleUI, parent: AppNode, i: number) {
  if (ui == null) {
    parent.children[i] = null;
  } else {
    const child = prepareChildNode(ui, parent, i);
    forceBuild(child);
  }
}

// Returns how many items it traversed, recursively.
// Because the ui array could be deeply nested, we need to flatten it. But
// doing that on every rebuild is costly. So this function does the same thing
// without the actual flattening.
//
// Because this function could be called recursively while traversing the
// AppNode's children, we need to tell it which child it should start at.
// TODO: handle keys.
function buildMultiUI(multiUI: UI[], parent: AppNode, startAt: number): number {
  let traversed = 0;
  for (let i = 0; i < multiUI.length; i++) {
    const r = multiUI[i];
    if (Array.isArray(r)) {
      // Recursively traverse the nested arrays and keep track of the traversal
      // count.
      traversed += buildMultiUI(r, parent, startAt + traversed);
    } else {
      buildUIAtIndex(r, parent, startAt + i);
      traversed++;
    }
  }
  return traversed;
}

function prepareChildNode(render: Render, parent: AppNode, i: number): AppNode {
  const node = parent.children[i];
  // TODO[perf]: If (node.render === render)
  //             skip building the subtree altogether.
  //             Does this optimization work for generators that need to be
  //             invoked multiple times?
  if (node && isSameComponentType(node.render, render)) {
    // In this case, we should reuse the existing app node.
    node.render = render;
    if (node.commitNode) {
      node.commitNode.dirtyState = DirtyState.SelfAndChildren;
    }
    return node;
  }
  return (parent.children[i] = new AppNode(parent, render));
}

function isSameComponentType(r1: Render, r2: Render) {
  // If both are null we shouldn't consider them of equal types.
  return r1.$$type && r2.$$type && r1.$$type == r2.$$type;
}

// Light-weight tree that maintains the full structure of the rendered component
// tree.
/*

CommitTree        AppTree
    o     <---       o
   / \             /   \
      o   <---    o     o
 / \             / \   / \
o         <---  o   o o   o

*/
export class AppNode<C extends CommitNode = CommitNode> {
  constructor(parent: AppNode | null, render: Render) {
    this.parent = parent;
    this.render = render;
  }

  dirtyState: DirtyState = DirtyState.SelfAndChildren;

  parent: AppNode | null;

  // Some children could be null when their corresponding UI materializes
  // to null. This guarantees that all state is blown away correctly.
  children: (AppNode | null)[] = [];

  // Commit nodes are only required when the component wants to have custom
  // commit logic (e.g. html elements).
  commitNode: C | null = null;

  state: any = null;

  render: Render;

  // Rebuild the sub-tree under this node.
  // TODO: This should accept a callback for updating state (ala React setState).
  // TODO: Make `rebuild` invalid during a build.
  rebuild() {
    this.markAsDirty();
    const rootNode = this.markParentChainAsDirty();
    rootNode.scheduleRebuild();
  }

  alreadyScheduledRebuild = false;

  // Should only be called on the root node.
  scheduleRebuild() {
    if (this.parent != null) {
      throw new Error(
        'scheduleRebuild() should only be called on the root node',
      );
    }
    if (!this.alreadyScheduledRebuild) {
      this.alreadyScheduledRebuild = true;
      requestAnimationFrame(() => {
        this.alreadyScheduledRebuild = false;
        build(this);
        // TODO[types]: This is the root node which is always committable.
        commit(this as CommittableAppNode);
      });
    }
  }

  protected markAsDirty() {
    this.dirtyState = DirtyState.SelfAndChildren;
    if (this.commitNode) {
      this.commitNode.dirtyState = DirtyState.SelfAndChildren;
    }
  }

  // Marks all nodes in the parent chain as dirty and returns the root app node.
  protected markParentChainAsDirty(): AppNode {
    // TODO[perf]: Maybe stop climbing as soon as we find a node that has a
    //             non-Clean dirty state.
    let node: AppNode = this;
    while (node.parent) {
      node = node.parent;
      if (node.dirtyState == DirtyState.Clean) {
        node.dirtyState = DirtyState.ChildrenOnly;
      }
      if (node.commitNode && node.commitNode.dirtyState == DirtyState.Clean) {
        node.commitNode.dirtyState = DirtyState.ChildrenOnly;
      }
    }
    return node;
  }
}

export type CommittableAppNode<C extends CommitNode = CommitNode> = AppNode<
  C
> & {
  commitNode: C;
};

export function isCommittable<C extends CommitNode = CommitNode>(
  node: AppNode<C>,
): node is CommittableAppNode<C> {
  return node.commitNode != null;
}
