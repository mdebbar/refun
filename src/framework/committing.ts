import { DirtyState } from './basic';
import { CommittableAppNode, AppNode, isCommittable } from './building';

export function commit(node: CommittableAppNode): any {
  if (node.commitNode.dirtyState !== DirtyState.Clean) {
    const children: any[] = [];
    commitChildren(node, children);
    node.commitNode.lastCommit = node.commitNode.commitWithChildren(children);
    node.commitNode.dirtyState = DirtyState.Clean;
  }
  return node.commitNode.lastCommit;
}

// Commits the children of `node` recursively. As it walks down the tree, it'll
// add children to `ancestorCommit`.
//
// QUESTION: should we allow commit nodes to throw? e.g. when adding children to
//           an <img>.
function commitChildren(node: AppNode, arr: any[]) {
  for (let i = 0; i < node.children.length; i++) {
    const child = node.children[i];
    if (child != null) {
      if (isCommittable(child)) {
        // When a child is committable, we commit it recursively and add its
        // commit as a child to the ancestor commit.
        arr.push(commit(child));
      } else {
        // Continue recursively in the tree to find more committable nodes.
        commitChildren(child, arr);
      }
    }
  }
}

// A `Node` in the commit tree. It knows how to commit itself during the commit
// phase after the build is complete. In the case of a browser, committing means
// updating html attributes, appending children, applying styles, etc.
export abstract class CommitNode<T = any> {
  dirtyState: DirtyState = DirtyState.SelfAndChildren;

  lastCommit: T;

  abstract getChildrenOfCommit(commit: T): T[];

  protected markAsDirty() {
    this.dirtyState = DirtyState.SelfAndChildren;
  }

  // Generally, this method shouldn't be overriden. Instead, override `commit`,
  // `appendChild`, `replaceChildAt`, `removeChildAt` and `trimChildren`.
  commitWithChildren(children: T[]): T {
    const commit =
      this.dirtyState === DirtyState.SelfAndChildren
        ? this.commit(this.lastCommit)
        : this.lastCommit;

    // This is an ugly HACK! Get rid of it when you find a better approach.
    const lastChildren =
      this.lastCommit && this.getChildrenOfCommit(this.lastCommit);

    // Diff children.
    const lastLength = lastChildren ? lastChildren.length : 0;
    for (let i = 0; i < children.length; i++) {
      if (i >= lastLength) {
        this.appendChild(commit, children[i]);
      } else if (children[i] !== lastChildren[i]) {
        // TODO[perf]: explore smarter ways to reuse children.
        this.replaceChildAt(commit, children[i], i);
      }
    }
    if (lastLength > children.length) {
      this.trimChildren(commit, lastChildren.length - children.length);
    }
    return commit;
  }

  abstract commit(lastCommit: T | null): T;

  abstract appendChild(commit: T, child: T): void;

  abstract replaceChildAt(commit: T, newChild: T, i: number): void;

  abstract removeChildAt(commit: T, i: number): void;

  abstract trimChildren(commit: T, count: number): void;
}
