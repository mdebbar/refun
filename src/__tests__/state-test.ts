import { run, component, UI, myStateNode } from '../framework';
import { NoopCommitter } from '../framework/commit';
import { stateSerializer } from '../test-helpers/snapshot-serializers';
import { findDescendantOfType } from '../framework/finders';

expect.addSnapshotSerializer(stateSerializer);

const app = (children: UI) => container(children);
const container = component(function container(children: UI) {
  myStateNode(null);
  return children;
});

test('simple build tree', () => {
  const root = run(() => app(null), new NoopCommitter());
  expect(root).toMatchSnapshot();
});

test('nested arrays', () => {
  const ui1 = () => app(null);
  const ui2 = () => app([[null]]);

  // The 2 snapshots should be identical.
  expect(run(ui1, new NoopCommitter())).toMatchSnapshot();
  expect(run(ui2, new NoopCommitter())).toMatchSnapshot();
});

test('complex arrays', () => {
  const dup = (children: UI) => [children, children];
  const ui = () => dup(dup([null, dup(null)]));
  const root = run(ui, new NoopCommitter());
  expect(root).toMatchSnapshot();
});

test.skip('stateless root component', () => {
  const parent = () => child();
  const child = component(function child() {
    const node = myStateNode(0);
    node.state++;
    return null;
  });

  const rootNode = run(parent(), new NoopCommitter());
  const parentNode = findDescendantOfType(rootNode, parent);
  expect(parentNode.state).toBe(1);

  rootNode.rebuild();
  expect(parentNode.state).toBe(2);
});

test('preserves state when rebuilding', () => {
  const app = component(function app() {
    myStateNode(null);
    return counter();
  });

  const counter = component(function counter() {
    const node = myStateNode(0);
    node.state++;
    return doubleCounter();
  });

  const doubleCounter = component(function doubleCounter() {
    const node = myStateNode(0);
    node.state += 2;
    return null;
  });

  const root = run(() => app(), new NoopCommitter());
  root.scheduleNextFrame = cb => cb();

  const counterNode = findDescendantOfType(root, counter);
  const doubleCounterNode = findDescendantOfType(root, doubleCounter);

  expect(counterNode.state).toBe(1);
  expect(doubleCounterNode.state).toBe(2);

  // Rebuild from the root.
  root.rebuild();
  // App nodes should be preserved.
  expect(findDescendantOfType(root, counter)).toBe(counterNode);
  expect(findDescendantOfType(root, doubleCounter)).toBe(doubleCounterNode);
  // States should've updated.
  expect(counterNode.state).toBe(2);
  expect(doubleCounterNode.state).toBe(4);

  // Rebuild from the counter node.
  counterNode.rebuild();
  expect(counterNode.state).toBe(3);
  expect(doubleCounterNode.state).toBe(6);

  // Rebuild from the double counter node.
  doubleCounterNode.rebuild();
  expect(counterNode.state).toBe(3);
  expect(doubleCounterNode.state).toBe(8);
});

test('complex state', () => {
  const counter = component(function counter(cb: (x: number) => UI) {
    const node = myStateNode(0);
    node.state++;
    return cb(node.state);
  });

  const doubleCount = component(function doubleCount(count: number) {
    const node = myStateNode(0);
    node.state = count * 2;
    return null;
  });

  const root = run(() => app(counter(doubleCount)), new NoopCommitter());

  const counterNode = findDescendantOfType(root, counter);
  const doubleCountNode = findDescendantOfType(root, doubleCount);

  expect(counterNode.state).toBe(1);
  expect(doubleCountNode.state).toBe(2);

  root.scheduleNextFrame = cb => cb();

  // Rebuild from the root.
  root.rebuild();
  expect(counterNode.state).toBe(2);
  expect(doubleCountNode.state).toBe(4);

  // Rebuild from the doubleCount node should keep the same state.
  doubleCountNode.rebuild();
  expect(counterNode.state).toBe(2);
  expect(doubleCountNode.state).toBe(4);
});

test('parent vs owner', () => {
  const parent = component(function parent(children: UI) {
    myStateNode(null);
    return [counter(), children];
  });

  const counter = component(function counter() {
    const node = myStateNode(0);
    node.state++;
    return null;
  });

  const rootNode = run(() => app(parent(counter())), new NoopCommitter());
  rootNode.scheduleNextFrame = cb => cb();
  expect(rootNode).toMatchSnapshot();

  // The counter owned by `parent` should be rebuilt, but the one passed in as a
  // child won't.
  findDescendantOfType(rootNode, parent).rebuild();
  expect(rootNode).toMatchSnapshot();
});
