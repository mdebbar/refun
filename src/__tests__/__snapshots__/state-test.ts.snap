// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex arrays 1`] = `
%root%
null
[
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  null
]
`;

exports[`nested arrays 1`] = `
%root%
null
[
  %container%
  null
  [
    null
  ]
]
`;

exports[`nested arrays 2`] = `
%root%
null
[
  %container%
  null
  [
    null
  ]
]
`;

exports[`parent vs owner 1`] = `
%root%
null
[
  %container%
  null
  [
    %parent%
    null
    [
      %counter%
      1
      [
        null
      ],
      %counter%
      1
      [
        null
      ]
    ]
  ]
]
`;

exports[`parent vs owner 2`] = `
%root%
null
[
  %container%
  null
  [
    %parent%
    null
    [
      %counter%
      2
      [
        null
      ],
      %counter%
      1
      [
        null
      ]
    ]
  ]
]
`;

exports[`simple build tree 1`] = `
%root%
null
[
  %container%
  null
  [
    null
  ]
]
`;
